rfcs/45-stageless.md

```diff
@@ -0,0 +1,540 @@
+Note that **this does not insert new copies of a `flush_commands` system**: instead, it is merely used to verify that your schedule has been set up correctly according the specified constraint.
+
+### Run criteria
+
+While ordering constraints determine *when* a system will run, **run criteria** will determine *if* it will run at all.
+A run criteria is a special kind of system, which can read data from the `World` and returns a boolean value.
+If its output is `true`, the system it is attached to will run during this pass of the `Schedule`;
+if it is `false`, the system will be skipped.
+Systems that are skipped are considered completed for the purposes of ordering constraints.
+If for some reason run criteria are themselves skipped, they are considered to have returned `true`.
+
+Let's examine a few ways we can specify run criteria:
+
+```rust
+// This function can be used as a run criteria system,
+// because it only reads from the `World` and returns `bool`
+fn construction_timer_finished(timer: Res<ConstructionTimer>) -> bool {
+    timer.finished()
+}
+
+fn main(){
+    App::new()
+    .add_plugins(DefaultPlugins)
+    // We can add functions as run criteria
+    .add_system(update_construction_progress.run_if(construction_timer_finished))
+    // We can use closures for simple one-off run criteria, 
+    // which automatically fetch the appropriate data from the `World`
+    .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+    // The `run_if_resource_is` method is convenient syntactic sugar that generates a run criteria
+    // for when you want to check the value of a resource (commonly an enum)
+    .add_system(gravity.run_if_resource_equals(Gravity::Enabled))
+    // Run criteria can be attached to labels: a copy of the run criteria will be applied to each system with that label
+    .configure_label(GameLabel::Physics.run_if_resource_equals(Paused(false)))
+    .run();
+}
+```
+
+**When multiple run criteria are attached to the same system, the system will run if and only if all of those run criteria return true.**
+
+Run criteria are evaluated "just before" the system that is attached to is run.
+The state that they read from will always be "valid" when the system that they control is being run: this is important to avoid strange bugs caused by race conditions and non-deterministic system ordering.
+It is impossible, for example, for a game to be paused *after* the run criteria checked if the game was paused but before the system that relied on the game not being paused completes.
+In order to understand exactly what this means, we need to understand atomic ordering constraints.
+
+### Atomic ordering constraints
+
+**Atomic ordering constraints** are a form of ordering constraints that ensure that two systems are run directly after each other.
+At least, as far as "directly after" is a meaningful concept in a parallel, nonlinear ordering.

```

Comment:
Atomic ordering constraints need a lot of design work to see if it's feasible. This is going to be especially tricky if we need to take sync points into account too as commands don't make it clear what archetypes they could modify.